### DISCUSSION TITLE
[7주차] 네트워크 계층: 제어 평면 5.5~7

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#15] [Q&A]
7주차 네트워크 계층: 제어 평면 5.5~7에 해당하는 질의응답입니다.

**Comment by Starlight258** [comment]
# Q1
`SDN`의 약자를 풀어 설명하고 `SDN`의 4가지 특징에 대해 설명해주세요.

# Q2
네트워크 관리 프로토콜인 `SNMP`와 `NETCONF`의 차이점을 중점으로 비교해서 설명해주세요.

**Reply by Starlight258** [reply]
> # Q1
> `SDN`의 약자를 풀어 설명하고 `SDN`의 4가지 특징에 대해 설명해주세요.

- `SDN` : 'Software-Defined Networking', 소프트웨어 정의 네트워크

### 특징
-  **플로우 기반 포워딩** : `SDN`으로 제어되는 스위치들에서의 패킷 포워딩은 `전송 계층`, `네트워크 계층`, 또는 `링크 계층` 헤더의 **어떤 값들을 기반으로도 이루어질 수 있다**.
- **데이터 평면과 제어 평면의 분리** 
	- `데이터 평면` : **네트워크 스위치들로 구성**된다.
		- `플로우 테이블` 기반으로 비교와 실행을 수행한다.
	- `제어 평면` : **서버와 스위치들의 플로우 테이블을 결정, 관리하는 소프트웨어**로 이루어진다.
- **네트워크 제어 기능이 데이터 평면 스위치 외부에 존재** : 제어 평면은 네트워크 스위치로부터 멀리 떨어진 별도의 몇 개 서버에서 구성된다.
	- `SDN 컨트롤러`와 `네트워크 제어 응용들의 집합`으로 구성된다.
- **프로그램이 가능한 네트워크** : 네트워크 제어 응용을 통해 네트워크를 프로그램 할 수 있다.

> # Q2
> 네트워크 관리 프로토콜인 `SNMP`와 `NETCONF`의 차이점을 중점으로 비교해서 설명해주세요.
- 주요 용도
  - SNMP : 네트워크 모니터링, 간단한 구성 변경
  - NETCONF : 복잡한 네트워크 구성 관리
- 작동 방식
  - SNMP : 주로 Polling
  - NETCONF : RPC (원격 프로시저 호출)
- 프로토콜
  - SNMP : 주로 UDP
  - NETCONF : 주로 TCP

**Reply by H-sooyeon** [reply]
> Q1 SDN의 약자를 풀어 설명하고 SDN의 4가지 특징에 대해 설명해주세요.

SDN은 software-defined networking, 소프트웨어 정의 네트워킹의 약자이다.
SDN은 네트워크의 제어 계층을 데이터 계층과 분리하여 네트워크를 소프트웨어적으로 중아에서 제어하고 관리할 수 있는 아키텍처이다.
SND의 네 가지 특징은 다음과 같다.
1. 민첩성.
비즈니스 및 애플리케이션 요구 사항이 변화되면 관리자는 필요에 따라 네트워크 구성을 조정할 수 있다.
2. 중앙집중식 관리
SDN은 네트워크 지능을 통합한다. 따라서 네트워크 구성과 활동을 전체적으로 파악할 수 있다.
3. 프로그래밍 가능
자동화된 SDN 서비스를 통해 네트워크 기능을 직접 프로그래밍하고 네트워크 리소스를 신속하고 쉽게 구성할 수 있다.
4. 개방형 연결성
SDN은 개방형 포준을 기반으로하며, 이러한 표준을 통해 구현된다.
그러므로 SDN 네트워크 설계를 효율화하고 공급업체 중립적 아키텍처에서 일관적인 네트워킹을 제공한다.


> Q2 네트워크 관리 프로토콜인 SNMP와 NETCONF의 차이점을 중점으로 비교해서 설명해주세요.

모두 네트워크 장비를 관리하기 위한 프로토콜이다.
SNMP는 모니터링에 중점을 두며 네트워크 장비에서 데이터를 조회하거나 변경할 수 있는 기능을 제공한다.
또한, 폴링 방식으로 동작하여 관리 시스템이 주기적으로 네트워크 장비에서 데이터를 가져온다.
SNMP는 단순 텍스트 기반의 프로토콜이며 주로 숫자 또는 문자열 형태로 데이터를 전송한다.

NETCONF는 구성관리에 초점을 맞춘 프로토콜이다.
네트워크 장비의 설정을 더 세밀하게 관리하고, 구성 상태를 읽고 쓰는 기능을 제공한다.
트랜잭션 기반으로, 구성 변경이 실패하면 롤백 기능을 제공하여 안정적인 구성을 유지할 수 있다.
또한, XML 또는 JSON과 같은 구조화된 데이터 포맷을 사용하여 네트워크 장비와 상호작용한다.

**Comment by choiyoung69** [comment]
## Q1
SDN 컨트롤러의 기능은 크게 3가지 계층으로 분리됩니다. 각 계층의 특징을 설명해주세요.

## Q2
네트워크 관리자가 네트워크를 관리하기 위해 사용하는 세 가지 방법에 대해 간단히 설명해주세요

**Reply by Starlight258** [reply]
> ## Q1
> SDN 컨트롤러의 기능은 크게 3가지 계층으로 분리됩니다. 각 계층의 특징을 설명해주세요.
> 
1. `통신 계층` : SDN 컨트롤러와 제어받는 네트워크 장치들 사이의 통신을 수행한다.
2. `네트워크 전역 상태 관리 계층` :  SDN 제어 평면의 제어 결정을 위해서 컨트롤러가 `네트워크 호스트`와 `링크`, `스위치`, 그리고 `SDN`으로 제어되는 다른 장치들에 대한 최신 정보를 얻는 계층이다.
3. `네트워크 제어 응용 계층과의 인터페이스` : `컨트롤러`는 `노스바운드`(northbound) 인터페이스를 통해서 네트워크 제어 응용들과 상호 작용을 수행한다.

> ## Q2
> 네트워크 관리자가 네트워크를 관리하기 위해 사용하는 세 가지 방법에 대해 간단히 설명해주세요
- CLI : `네트워크 운영자`는 명령줄 인터페이스를 통해 **장치에 직접 명령을 보낸다**.
- SNMP/MIB : 네트워크 운영자는 `SNMP`를 사용하여 **장치의 MIB에 있는 데이터를 질의하거나 설정**한다.
- NETCONF/YANG 
  - `NETCONF` : 설정 관리에 초점을 둔 더 현대적인 프로토콜
  - `YANG` : `NETCONF`와 함께 사용되는 데이터 모델링 언어

**Reply by H-sooyeon** [reply]
> Q1 SDN 컨트롤러의 기능은 크게 3가지 계층으로 분리됩니다. 각 계층의 특징을 설명해주세요.

SDN 컨트롤러는 애플리케이션 계층, 제어 계층, 데이터 계층으로 나뉜다.
`애플리케이션 계층`은 SDN의 컨트롤러가 제공하는 네트워크 정보를 바탕으로 동작하며, 네트워크 정책, 트래픽 관리, 보안, 최적화 등의 기능을 수행한다.

`제어 계층`은 SDN 컨트롤러의 핵심 기능이 위치하는 부분으로, 네트워크 전체를 중앙에서 제어하고 관리하는 역할을 담당한다. 물리적 네트워크 장비(스위치, 라우터)들과 애플리케이션 계층간의 중재 역할을 한다.
또한 네트워크 토폴로지 정보를 수집하고, 이를 바탕으로 최적의 경로 설정, 트래픽 흐름 관리 등을 수행한다.

`데이터 계층`은 데이터 평면에서 동작하며 네트워크의 실제 데이터 패킷이 전달되는 물리적 또는 가상 네트워크 장비로 구성된다.
컨트롤러가 정의한 규칙에 따라 트래픽을 처리하고 전달한다.

> Q2 네트워크 관리자가 네트워크를 관리하기 위해 사용하는 세 가지 방법에 대해 간단히 설명해주세요

1. CLI
네트워크 장비에 직접 접속하여 명령어를 입력해 설정을 변경하고 상태를 모니터링하는 방법이다.
2. SNMP/MIB
네트워크 운영자는 SNMP로 네트워크 장비를 관리하고 모니터링하며 장비와 통신할 때 MIB를 사용해 데이터를 조회하거나 변경한다.
MIB는 각 네트워크 자입의 관리 정보를 정의한 데이터베이스이다.
3. NETCONF/YANG
NETCONF로 네트워크 장비의 구성을 관리 . 및변경하며 네트워크 장비의 설정을 프로그래밍 방식으로 제어한다.
또한 NETCONF는 YANG이라는 데이터 모델링 언어를 사용해 NETCONF가 네트워크 장비의 설정을 이해하고 관리할 수 있도록 한다.

**Comment by H-sooyeon** [comment]
## Q1

SDN의 장점에 대해 설명해주세요.


## Q2

SDN이 어떤 곳에 어떤식으로 활용되는지 예를 들어 설명해주세요.

**Reply by Starlight258** [reply]
> ## Q1
> SDN의 장점에 대해 설명해주세요.
> 
1. `플로우 기반 포워딩` : `SDN`으로 제어되는 스위치들의 패킷 포워딩은 `전송 계층`, `네트워크 계층`, 또는 `링크 계층` 헤더의 **어떤 값들을 기반으로도 이루어질 수 있다**.
> 더 유연하게 포워딩할 수 있다.

2. `데이터 평면과 제어 평면의 분리` : `데이터 평면`은 네트워크 스위치로 구성되며, `제어 평면`은 서버와 스위치들의 플로우 테이블을 결정하는 소프트웨어로 이루어진다.
> `최적의 로드밸런싱`이 가능하다.

3. **`프로그램이 가능한 네트워크`** : 네트워크 제어 응용을 통해 네트워크를 프로그램 할 수 있다.
> 운영의 효율성을 높여 비용이 절감된다.

<br/>

> ## Q2
> SDN이 어떤 곳에 어떤식으로 활용되는지 예를 들어 설명해주세요.

`구글의 G-Scale` : 구글 데이터센터 백본 구간을 SDN으로 전환 (2012)
 -  네트워크 인프라의 활용도를 4~50% 수준에서 100%로 끌어올렸다.
 - WAN 대역의 경로 최적화하여 사용자 경험을 향상시켰다.
 - 네트워크 구축 비용이 절감되었다.

**Comment by Hanjaemo** [comment]
### Q1
SDN이 등장하게 된 배경(이유)에 대해 설명해 주세요.

### Q2
SDN의 동작 원리를 데이터 평면과 제어 평면의 상호작용 측면에서 설명해 주세요.

**Reply by Starlight258** [reply]
> ### Q1
> SDN이 등장하게 된 배경(이유)에 대해 설명해 주세요.
> 
1. `네트워크 사용 환경의 변화`
- 스마트폰, 태블릿 PC 등 모바일 기기가 증가하였고 SNS가 대중화되었으며 클라우드 컴퓨팅, 빅데이터 등 새로운 IT 기술이 등장하였다.
- 데이터 트래픽이 급격하게 증가하였으며 네트워크에 대한 요구사항도 증가하였다. (속도, 안전성)
- 하지만 기존 하드웨어 중심의 구조는 유연성이 부족하고 네트워크 관리와 운영이 복잡하여 한계가 있었다.

2. `복잡한 네트워크 환경` 
- 벤더사마다 프로토콜이 달라 통합 관리하는 것이 어렵고, 각 장비마다 다른 사용자 인터페이스를 가지고 있었다.
- 다양한 네트워크와 인터페이스로 인해 전체 네트워크를 일관되게 운영하고 관리하는 것이 어려웠다.

=> 대량의 데이터를 처리할 수 있고 유동적으로 변화할 수 있는 네트워크 환경을 필요로 하여,  **소프트웨어로 쉽게 환경을 구성하고 운영할 수 있는 SDN이 등장**하게 되었다.

> ### Q2
> SDN의 동작 원리를 데이터 평면과 제어 평면의 상호작용 측면에서 설명해 주세요.

- `데이터 평면` : 플로우 테이블에 기반하여 패킷 포워딩 수행
- `제어 평면` : 데이터 평면 장치들의 플로우 테이블을 설정하고 관리
- 상호 작용
  - 1. `상태 보고` : 데이터 평면 장치가 `OpenFlow` 프로토콜을 통해 네트워크 상태 변화를 컨트롤러에 보고한다.
  - 2. `상태 업데이트` : 컨트롤러는 네트워크 전역 상태를 업데이트 한다.
  - 3. `경로 재계산` : 네트워크 제어 응용이 새로운 상태를 바탕으로 최적의 경로를 재계산한다.
  - 4. `플로우 테이블 갱신` : 컨트롤러가 `OpenFlow` 프로토콜을 통해 데이터 평면 장치들의 테이블을 갱신한다.

**Reply by H-sooyeon** [reply]
> Q1 SDN이 등장하게 된 배경(이유)에 대해 설명해 주세요.

SDN은 네트워크 구조의 한계를 극복하고자 등장한 새로운 네트워크 아키텍처이다.
기존 네트워크의 복잡성과 비효율성, 네트워크 확장성 관련 문제, 신기술 발전 등의 문제로 SDN이 등장하게 되었다.
SDN은 제어평면과 데이터 평면의 분리, 중앙 집중식 제어, 프로그램 가능성의 특징이 있다.

> Q2 SDN의 동작 원리를 데이터 평면과 제어 평면의 상호작용 측면에서 설명해 주세요.

제어 평면이 데이터 평면을 중아에서 제어하고, 각 장비는 컨트롤러의 지시에 따라 패킷을 처리하는 구조이다.
이를 통해 네트워크 유연성과 효율성을 높이고, 중앙에서 프로그래밍 가능한 네트워크 관리가 가능해진다.

네트워크에 새로운 호스트가 연결되고, 그 호스트에서 다른 네트워크로 패킷을 전송한다면, 아래와 같은 흐름으로 동작한다.

1. 패킷 전송: 호스트가 패킷을 전송하면, 스위치(데이터 평면)에 패킷이 도착한다.
2. 경로 질의: 스위치는 아직 해당 호스트와 목적지 간의 경로 정보를 가지고 있지 않으므로, 패킷을 SDN 컨트롤러로 전달하여 경로를 요청한다.
3. 경로 결정: SDN 컨트롤러는 네트워크 전체 상태를 바탕으로 최적의 경로를 계산하고, 스위치에게 해당 패킷을 어느 포트로 포워딩해야 할지 명령을 내린다.
4. 패킷 처리: 스위치는 컨트롤러의 명령에 따라 패킷을 처리하고, 지정된 경로로 전달한다.
5. 플로우 테이블 업데이트: 스위치는 컨트롤러가 제공한 경로와 정책을 플로우 테이블에 기록하여, 동일한 트래픽이 발생할 때 신속히 처리할 수 있도록 준비한다.
